# Pandas Tutorial

Pandas is a Python library.

Pandas is used to analyze data.

Starting with a basic introduction and ends up with cleaning and plotting data:

Basic
 Introduction,
 Getting Started,
 pandas series,
 DataFrames,
 Read CSV,
 Read Json,
 Analyze Data,

Cleaning Data
 Clean Data,
 Clean Empty Cells,
 Clean Wrong Format,
 Clean Wrong Data,
 Remove Duplicates,
 
Advanced
 Corelations,
 Plotting

# What is Pandas?
Pandas is a Python library used for working with data sets.

It has functions for analyzing, cleaning, exploring, and manipulating data.

The name "Pandas" has a reference to both "Panel Data", and "Python Data Analysis" and was created by Wes McKinney in 2008.

# Why Use Pandas?
Pandas allows us to analyze big data and make conclusions based on statistical theories.

Pandas can clean messy data sets, and make them readable and relevant.

Relevant data is very important in data science.

:}Data Science: is a branch of computer science where we study how to store, use and analyze data for deriving information from it.

# What Can Pandas Do?
Pandas gives you answers about the data. Like:

Is there a correlation between two or more columns?

What is average value?

Max value?

Min value?

Pandas are also able to delete rows that are not relevant, or contains wrong values, like empty or NULL values. This is called cleaning the data.



In [3]:
import pandas

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}
myvar = pandas.DataFrame(mydataset)
print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


# Pandas Series
# What is a Series?
A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

In [8]:
# Create a simple Pandas Series from a list:
import pandas as pd
a = [5,3,8,7]
ser = pd.Series(a)
print(ser)

0    5
1    3
2    8
3    7
dtype: int64


# Labels
If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

This label can be used to access a specified value.

In [9]:
# Return the first value of the Series:
print(a[0])

5


# Create Labels
With the index argument, you can name your own labels.

In [10]:
a = [5,3,8,7]
ser = pd.Series(a, index=['x','y','z','h'])
print(ser)

x    5
y    3
z    8
h    7
dtype: int64


When you have created labels, you can access an item by referring to the label.

In [12]:
# return the value of z
print(ser['z'])

8


# Key/Value Objects as Series
You can also use a key/value object, like a dictionary, when creating a Series.

In [13]:
calories = {"day1":420, "day2":380, "day3":390}
x = pd.Series(calories)
print(x)

day1    420
day2    380
day3    390
dtype: int64


Note: The keys of the dictionary become the labels.

To select only some of the items in the dictionary, use the index argument and specify only the items you want to include in the Series.

In [15]:
# Create a Series using only data from "day1" and "day2":
calories = {"day1":420, "day2":380, "day3":390}
x = pd.Series(calories, index=["day1", "day2"])
print(x)

day1    420
day2    380
dtype: int64


# DataFrames
Data sets in Pandas are usually multi-dimensional tables, called DataFrames.

Series is like a column, a DataFrame is the whole table.

In [16]:
# Create a DataFrame from two Series:
data = {"calories":[420,380,390],
        "duration":[50,45,90]}
x = pd.DataFrame(data)
print(x)

   calories  duration
0       420        50
1       380        45
2       390        90


# Pandas DataFrames
# What is a DataFrame?
A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

In [17]:
# Create a simple Pandas DataFrame:
data = {"calories":[420,380,390],
        "duration":[50,45,90]}
# load data into dataframe object
df = pd.DataFrame(data)
print(df)

   calories  duration
0       420        50
1       380        45
2       390        90


# Locate Row
As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the loc attribute to return one or more specified row(s)

In [18]:
#Return row 0:

#refer to the row index:
print(df.loc[0])


calories    420
duration     50
Name: 0, dtype: int64


Note: This example returns a Pandas Series.

In [21]:
# Return row 0 and 1:
print(df.loc[[0, 1]])

   calories  duration
0       420        50
1       380        45


Note: When using [], the result is a Pandas DataFrame.

# Named Indexes
With the index argument, you can name your own indexes.

In [23]:
#Add a list of names to give each row a name:
data = {"calories":[420,380,390],
        "duration":[50,45,90]}
df = pd.DataFrame(data, index=["day1", "day2", "day3"])
print(df)

      calories  duration
day1       420        50
day2       380        45
day3       390        90


# Locate Named Indexes
Use the named index in the loc attribute to return the specified row(s).

In [24]:
# Return "day2":

#refer to the named index:
print(df.loc["day1"])

calories    420
duration     50
Name: day1, dtype: int64


# Load Files Into a DataFrame
If your data sets are stored in a file, Pandas can load them into a DataFrame.

In [26]:
# Load a comma separated file (CSV file) into a DataFrame:
df = pd.read_csv('data.csv')
print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


# Pandas Read CSV
# Read CSV Files
A simple way to store big data sets is to use CSV files (comma separated files).

CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

In our examples we will be using a CSV file called 'data.csv'.

In [27]:
#Load the CSV into a DataFrame:
df = pd.read_csv('data.csv')
print(df.to_string())
# Tip: use to_string() to print the entire DataFrame.

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

If you have a large DataFrame with many rows, Pandas will only return the first 5 rows, and the last 5 rows:

In [28]:
# Print the DataFrame without the to_string() method:
df = pd.read_csv('data.csv')
print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


# max_rows
The number of rows returned is defined in Pandas option settings.

You can check your system's maximum rows with the pd.options.display.max_rows statement.

In [29]:
df = pd.read_csv('data.csv')
print(pd.options.display.max_rows)

60


the number is 60, which means that if the DataFrame contains more than 60 rows, the print(df) statement will return only the headers and the first and last 5 rows.

You can change the maximum rows number with the same statement.

In [30]:
# Increase the maximum number of rows to display the entire DataFrame:
pd.options.display.max_rows = 9999
df = pd.read_csv('data.csv')
print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

# Pandas - Analyzing DataFrames
# Viewing the Data
One of the most used method for getting a quick overview of the DataFrame, is the head() method.

The head() method returns the headers and a specified number of rows, starting from the top.

In [32]:
# Get a quick overview by printing the first 10 rows of the DataFrame:
df = pd.read_csv('data.csv')
print(df.head(10))
# Note: if the number of rows is not specified, the head() method will return the top 5 rows.

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


In [34]:
# Print the first 5 rows of the DataFrame:
df = pd.read_csv('data.csv')
print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


There is also a tail() method for viewing the last rows of the DataFrame.

The tail() method returns the headers and a specified number of rows, starting from the bottom.


In [35]:
# Print the last 5 rows of the DataFrame:

print(df.tail()) 

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


# Info About the Data
The DataFrames object has a method called info(), that gives you more information about the data set.

In [36]:
#Print information about the data:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


# Result Explained
The result tells us there are 169 rows and 4 columns:


  RangeIndex: 169 entries, 0 to 168
  Data columns (total 4 columns):

And the name of each column, with the data type:


   #   Column    Non-Null Count  Dtype  
  ---  ------    --------------  -----  
   0   Duration  169 non-null    int64  
   1   Pulse     169 non-null    int64  
   2   Maxpulse  169 non-null    int64  
   3   Calories  164 non-null    float64

# Null Values
The info() method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.

Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. This is a step towards what is called cleaning data, and you will learn more about that in the next chapters.

# Pandas - Cleaning Data
# Data Cleaning
Data cleaning means fixing bad data in your data set.

Bad data could be:

Empty cells
Data in wrong format
Wrong data
Duplicates
In this tutorial you will learn how to deal with all of them.

# Our Data Set
In the next chapters we will use this data set:


      Duration          Date  Pulse  Maxpulse  Calories
  0         60  '2020/12/01'    110       130     409.1
  1         60  '2020/12/02'    117       145     479.0
  2         60  '2020/12/03'    103       135     340.0
  3         45  '2020/12/04'    109       175     282.4
  4         45  '2020/12/05'    117       148     406.0
  5         60  '2020/12/06'    102       127     300.0
  6         60  '2020/12/07'    110       136     374.0
  7        450  '2020/12/08'    104       134     253.3
  8         30  '2020/12/09'    109       133     195.1
  9         60  '2020/12/10'     98       124     269.0
  10        60  '2020/12/11'    103       147     329.3
  11        60  '2020/12/12'    100       120     250.7
  12        60  '2020/12/12'    100       120     250.7
  13        60  '2020/12/13'    106       128     345.3
  14        60  '2020/12/14'    104       132     379.3
  15        60  '2020/12/15'     98       123     275.0
  16        60  '2020/12/16'     98       120     215.2
  17        60  '2020/12/17'    100       120     300.0
  18        45  '2020/12/18'     90       112       NaN
  19        60  '2020/12/19'    103       123     323.0
  20        45  '2020/12/20'     97       125     243.0
  21        60  '2020/12/21'    108       131     364.2
  22        45           NaN    100       119     282.0
  23        60  '2020/12/23'    130       101     300.0
  24        45  '2020/12/24'    105       132     246.0
  25        60  '2020/12/25'    102       126     334.5
  26        60    2020/12/26    100       120     250.0
  27        60  '2020/12/27'     92       118     241.0
  28        60  '2020/12/28'    103       132       NaN
  29        60  '2020/12/29'    100       132     280.0
  30        60  '2020/12/30'    102       129     380.3
  31        60  '2020/12/31'     92       115     243.0

The data set contains some empty cells ("Date" in row 22, and "Calories" in row 18 and 28).

The data set contains wrong format ("Date" in row 26).

The data set contains wrong data ("Duration" in row 7).

The data set contains duplicates (row 11 and 12).

# Pandas - Cleaning Empty Cells
# Empty Cells
Empty cells can potentially give you a wrong result when you analyze data.

# Remove Rows
One way to deal with empty cells is to remove rows that contain empty cells.

This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.

In [13]:
# Return a new Data Frame with no empty cells:
df = pd.read_csv('data.csv')

new_df = df.dropna()

print(new_df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

Note: By default, the dropna() method returns a new DataFrame, and will not change the original.

If you want to change the original DataFrame, use the inplace = True argument:

In [14]:
# Remove all rows with NULL values:

import pandas as pd

df = pd.read_csv('data.csv')

df.dropna(inplace = True)

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

Note: Now, the dropna(inplace = True) will NOT return a new DataFrame, but it will remove all rows containing NULL values from the original DataFrame.

# Replace Empty Values
Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value:

In [15]:
# Replace NULL values with the number 130:

import pandas as pd

df = pd.read_csv('data.csv')

df.fillna(130, inplace = True)

# Replace Only For Specified Columns
The example above replaces all empty cells in the whole Data Frame.

To only replace empty values for one column, specify the column name for the DataFrame:

In [17]:
# Replace NULL values in the "Calories" columns with the number 130:
df['Calories'].fillna(130,inplace=True)
#This operation inserts 130 in empty cells in the "Calories" column (row 18 and 28).

# Replace Using Mean, Median, or Mode
A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

#Calculate the MEAN, and replace any empty values with it:

In [20]:

pd.read_csv('data.csv')
x = df['Calories'].mean()
df['Calories'].fillna(x, inplace=True)
# Mean = the average value (the sum of all values divided by number of values).

In [23]:
# Calculate the MEDIAN, and replace any empty values with it:
df = pd.read_csv('data.csv')
x = df['Calories'].median()
df['Calories'].fillna(x,inplace=True)
# Median = the value in the middle, after you have sorted all values ascending.

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     318.6
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [27]:
# Calculate the MODE, and replace any empty values with it:
df = pd.read_csv('data.csv')
x = df['Calories'].mode()[0]
df['Calories'].fillna(x,inplace=True)
# Mode = the value that appears most frequently.